### PR TITLE
Rework region mechanics

### PR DESCRIPTION
**Pull request by Pablete1234**
If a region is bounded or unbounded should not be decided by it's class, it should be defined by if the boundries are finite or not, it will help implementing region transformations in the future

### CODE REVIEW

File: src/main/java/in/twizmwaz/cardinal/module/region/type/BlockBoundedRegion.java [hunk 1]
```diff
@@ -26,15 +26,54 @@
...
+
+  @Override
+  public Vector getRandomPoint() {
+    return blocks.get(new Random().nextInt(blocks.size() + 1)).getLocation().toVector();
+  }
+
+  @Override
+  public boolean evaluate(Vector vector) {
+    for (Block block : blocks) {
+      if (block.getLocation().toVector().equals(vector)) {
```
Comment (iEli2tyree011) [inline src/main/java/in/twizmwaz/cardinal/module/region/type/BlockBoundedRegion.java]: What if the vector is in between blocks?

Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/region/type/BlockBoundedRegion.java]: that's supposed to send evaluate to it's parent region, woops

File: src/main/java/in/twizmwaz/cardinal/module/region/RegionBounds.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  public Vector getCenter() {
+    return min.midpoint(max);
+  }
+
+  public BlockRegion getCenterBlock() {
+    return new BlockRegion(getCenter());
+  }
+
+  public BlockVector blockAlign(Vector vector) {
+    return new BlockVector((int) vector.getX() + 0.5d, (int) vector.getY() + 0.5d, (int) vector.getZ() + 0.5d);
```
Comment (iEli2tyree011) [inline src/main/java/in/twizmwaz/cardinal/module/region/RegionBounds.java]: Casting getX() to an integer is the same thing as using getBlockX()

File: src/main/java/in/twizmwaz/cardinal/module/region/RegionBounds.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+import org.bukkit.util.Vector;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Data
+@AllArgsConstructor
+public class RegionBounds {
+
+  Vector min, max;
```
Comment (iEli2tyree011) [inline src/main/java/in/twizmwaz/cardinal/module/region/RegionBounds.java]: These fields should have a defined scope

File: src/main/java/in/twizmwaz/cardinal/module/region/RegionBounds.java (hunk 2 above)
Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/region/RegionBounds.java]: From experience when doing the RegionSaves in cardinal1, adding 0.5 is better

Comment (iEli2tyree011) [inline src/main/java/in/twizmwaz/cardinal/module/region/RegionBounds.java]: That's not my point, you don't need to cast vector.getX() to an integer. Just use vector.getBlockX().

Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/region/RegionBounds.java]: i changed it

File: src/main/java/in/twizmwaz/cardinal/util/Vectors.java [hunk 4]
```diff
@@ -66,4 +66,18 @@ public static Vector getVector(String str) {
...
+  }
+
+  public static Vector max() {
+    return new Vector(Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY);
+  }
+
+  public static Vector getMirroredVector(Vector vector, Vector origin, Vector normal) {
+    vector = vector.minus(origin);
+    Vector projected = vector.minus(normal.times(vector.getCrossProduct(normal).multiply(2)));
+    return projected.add(origin);
```
Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/util/Vectors.java]: I'm not sure if this function actually works, but i think it should, any good way of testing this?

Comment (iEli2tyree011) [inline src/main/java/in/twizmwaz/cardinal/util/Vectors.java]: Make it log the original and new vector, and check if it's the right values.

File: src/main/java/in/twizmwaz/cardinal/module/region/type/modifications/TranslatedRegion.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    this(region, offset, region.getBounds().translate(offset));
+  }
+
+  @Override
+  public boolean evaluate(Vector vector) {
+    return region.evaluate(vector.minus(offset));
+  }
+
+  @Override
+  public RegionBounds getBounds() {
```
Comment (twizmwazin) [inline src/main/java/in/twizmwaz/cardinal/module/region/type/modifications/TranslatedRegion.java]: Should use lombok for this

File: src/main/java/in/twizmwaz/cardinal/util/Vectors.java [hunk 6]
```diff
@@ -66,4 +66,23 @@ public static Vector getVector(String str) {
...
+
+  public static Vector getMirroredVector(Vector vector, Vector origin, Vector normal) {
+    vector = vector.minus(origin);
+    vector = vector.minus(normal.times(vector.dot(normal)).times(2)).add(origin);
+    vector = new Vector(round(vector.getX()), round(vector.getY()), round(vector.getZ()));
+    return vector;
+  }
+
+  public static double round(double d) {
+    return (double)Math.round(d * 10) / 10D;
```
Comment (twizmwazin) [inline src/main/java/in/twizmwaz/cardinal/util/Vectors.java]: Don't know if checkstyle picked up on this, but you're missing a space

File: src/main/java/in/twizmwaz/cardinal/module/region/type/modifications/MirroredRegion.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    this(region, origin, normal.normalize(), region.getBounds().mirror(origin, normal.normalize()));
+  }
+
+  @Override
+  public boolean evaluate(Vector vector) {
+    return region.evaluate(Vectors.getMirroredVector(vector, origin, normal));
+  }
+
+  @Override
+  public RegionBounds getBounds() {
```
Comment (twizmwazin) [inline src/main/java/in/twizmwaz/cardinal/module/region/type/modifications/MirroredRegion.java]: lombok

File: src/main/java/in/twizmwaz/cardinal/module/region/type/SphereRegion.java [hunk 8]
```diff
@@ -23,43 +23,47 @@
...
-  public BlockRegion getCenterBlock() {
-    return new BlockRegion(origin);
+  public boolean evaluate(Vector evaluating) {
+    return evaluating.isInSphere(origin, radius);
   }
 
   @Override
-  public boolean evaluate(Vector evaluating) {
-    return evaluating.isInSphere(origin, radius);
+  public RegionBounds getBounds() {
```
Comment (twizmwazin) [inline src/main/java/in/twizmwaz/cardinal/module/region/type/SphereRegion.java]: lombok

File: src/main/java/in/twizmwaz/cardinal/module/region/type/modifications/MirroredRegion.java (hunk 7 above)
Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/region/type/modifications/MirroredRegion.java]: an @Getter should do it, right?

Comment (twizmwazin) [inline src/main/java/in/twizmwaz/cardinal/module/region/type/modifications/MirroredRegion.java]: yup

**Pull request closed without merging**